File: Spring/hello-spring/src/main/java/polymorphism/SamsungTV.java
package polymorphism;

import org.springframework.stereotype.Component;

@Component
public class SamsungTV implements TV {
    private Speaker speaker;
    private int price;

    public void initMethod() {
        System.out.println("객체 초기화 작업 처리....");
    }

    public void destroyMethod() {
        System.out.println("객체 삭제 전에 처리할 로직 처리...");
    }
    public SamsungTV() {
        System.out.println("SamsungTV(1) -- 객체 생성");
    }

    public SamsungTV(Speaker speaker, int price) {
        System.out.println("SamsungTV(2) -- 객체 생성");
        this.speaker = speaker;
        this.price = price;
    }
    public void powerOn() {
        System.out.printf("SamsungTV -- Power On\t(가격: %d)\n", price);
    }

    public void powerOff() {
        System.out.println("SamsungTV -- Power Off");
    }


    public void volumeUp() {
        speaker.volumeUp();
    }

    public void volumeDown() {
        speaker.volumeDown();
    }
}

File: kotlin_spring/src/main/kotlin/com/example/kotlin_spring/polymorphism/tv/SamsungTV.kt
package com.example.kotlin_spring.polymorphism.tv

import com.example.kotlin_spring.polymorphism.speaker.Speaker
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import kotlin.properties.Delegates

class SamsungTV(private val speaker: Speaker, private val price: Int = 2500000) : TV {
    fun initMethod() {
        println("SamsungTV -- Init")
    }

    fun destroyMethod() {
        println("SamsungTV -- Destroy")
    }

    init {
        println("SamsungTV 객체 생성, price = $price")
    }

    override fun powerOn() {
        println("SamsungTV --- On")
    }

    override fun powerOff() {
        println("SamsungTV --- Off")
    }

    override fun volumeUp() {
        speaker.volumeUp()
    }

    override fun volumeDown() {
        speaker.volumeDown()
    }
}